### PR TITLE
Carry out mapping using the NCI Metathesaurus

### PR DESCRIPTION
**Pull request by gaurav**
A simple term mapper built around the NCI Metathesaurus. It can produce the entire known mapping, or only map a list of input terms. Most RRF files are loaded directly into memory, but since two critical files we need (MRCONSO and MRHIER) are particularly large, we load those into a local SQLite3 database and query them from there. It also includes support for matching via the parent term, i.e. when term X in source A has no mapping in source B, but X's parent term does.

### CODE REVIEW

File: src/main/scala/org/renci/umls/db/DbConcepts.scala [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    while(rs.next()) {
+      results = (rs.getString(1), rs.getString(2)) +: results
+    }
+    conn.close()
+
+    results.groupMap(_._1)(_._2)
+  }
+}
+
+/** Represents a single column entry. */
```
Comment (gaurav) [inline src/main/scala/org/renci/umls/db/DbConcepts.scala]: Just realized that this duplicates the Concept in RRFConcepts, so I'll remove this copy.

File: src/main/scala/org/renci/umls/CodeMapper.scala [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    // Both sourceFrom and sourceTo are set!
+    if (conf.idFile.isEmpty) {
+      val maps = concepts.getMap(conf.fromSource(), Seq.empty, conf.toSource(), Seq.empty)
+      stream.println("fromSource\tfromCode\ttoSource\ttoCode\tnciMTConceptIds\tlabels")
+      maps.foreach(map => {
+        stream.println(
+          s"${map.fromSource}\t${map.fromCode}\t" +
+          s"${map.toSource}\t${map.toCode}\t" +
+          s"${map.conceptIds.mkString(", ")}\t" +
+          s"${map.labels}"
```
Comment (gaurav) [inline src/main/scala/org/renci/umls/CodeMapper.scala]: This should have a `mkString("|")` for the labels.

File: src/main/scala/org/renci/umls/db/DbConcepts.scala (hunk 1 above)
Comment (gaurav) [reply src/main/scala/org/renci/umls/db/DbConcepts.scala]: Fixed in 95f6ff6.

File: src/main/scala/org/renci/umls/CodeMapper.scala (hunk 2 above)
Comment (gaurav) [reply src/main/scala/org/renci/umls/CodeMapper.scala]: Fixed in 0ad195b.

**Comment by gaurav** [comment]
Sorry for the post-review request changes, Jim: I fixed a couple of minor issues and added CUIs to the CSV output. This PR should be stable enough for you to review now!

File: build.sbt [hunk 3]
```diff
@@ -0,0 +1,69 @@
+// Publication information
+name := "umls-rrf"
+ThisBuild / organization := "com.ggvaidya"
+ThisBuild / version      := "0.1-SNAPSHOT"
```
Comment (balhoff) [inline build.sbt]: Is `ThisBuild` necessary? I usually just have `organization` and `version`.

File: build.sbt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  "scalafixCheckAll",
+  "; compile:scalafix --check ; test:scalafix --check"
+)
+
+// Library dependencies.
+libraryDependencies ++= {
+  Seq(
+    // Logging
+    "com.typesafe.scala-logging"  %% "scala-logging"          % "3.9.2",
+    "ch.qos.logback"              %  "logback-classic"        % "1.2.3",
```
Comment (balhoff) [inline build.sbt:38]: This is just a comment; I've been liking the extreme simplicity of `scribe`: `"com.outr" %% "scribe-slf4j" % "2.7.12"`.

File: src/main/scala/org/renci/umls/db/DbConcepts.scala [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+import scalacache._
+import scalacache.caffeine._
+import scalacache.memoization._
+import scalacache.modes.sync._
+
+import scala.concurrent.duration._
+import scala.collection.mutable
+import scala.io.Source
+
+/** A wrapper for RRFConcepts that uses  */
```
Comment (balhoff) [inline src/main/scala/org/renci/umls/db/DbConcepts.scala]: that uses...? :-)

File: src/main/scala/org/renci/umls/db/DbConcepts.scala [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+  implicit val halfMapCache: Cache[Seq[HalfMap]] = CaffeineCache[Seq[HalfMap]]
+
+  /** The name of the table used to store this information. We include the SHA-256 hash so we reload it if it changes. */
+  val tableName: String = "MRCONSO_" + sha256
+
+  /* Check to see if the MRCONSO_ table seems up to date. If not, load it into memory from the file. */
+  val conn1 = db.createConnection()
+  val checkCount = conn1.createStatement()
+  val results = Try { checkCount.executeQuery(s"SELECT COUNT(*) AS cnt FROM $tableName") }
+  val rowsFromDb = if (results.isSuccess) results.get.getInt(1) else -1
```
Comment (balhoff) [inline src/main/scala/org/renci/umls/db/DbConcepts.scala]: I prefer `val rowsFromDb = results.map(_.getInt(1)).getOrElse(-1)`

**Review by balhoff — APPROVED** [review]
I'm curious, did you look at any of the Scala database libraries, like Doobie or Slick?

File: build.sbt (hunk 4 above)
Comment (gaurav) [reply build.sbt:38]: Ooo, nice! I'll check that out (https://github.com/gaurav/umls-rrf-scala/issues/4).

File: build.sbt (hunk 3 above)
Comment (gaurav) [reply build.sbt]: Ha! I'm not sure where I got that from -- it looks like it's only needed in [multiproject builds](https://www.scala-sbt.org/1.x/docs/Scopes.html#Scoping+by+the+subproject+axis). Removed in c5cbd11.

File: src/main/scala/org/renci/umls/db/DbConcepts.scala (hunk 5 above)
Comment (gaurav) [reply src/main/scala/org/renci/umls/db/DbConcepts.scala]: D'oh! Fixed in 41a26bb.

File: src/main/scala/org/renci/umls/db/DbConcepts.scala (hunk 6 above)
Comment (gaurav) [reply src/main/scala/org/renci/umls/db/DbConcepts.scala]: That is much cleaner! Fixed in 7995c55.

**Pull request merged**